src/Microsoft.Data.Analysis/PrimitiveColumnContainer.cs

```diff
@@ -190,64 +198,69 @@ public void AppendMany(T? value, long count)
-                    Buffers.Add(lastBuffer);
+                    Buffers.Add(new DataFrameBuffer<T>());
                     NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
                 }
-                DataFrameBuffer<T> mutableLastBuffer = DataFrameBuffer<T>.GetMutableBuffer(lastBuffer);
-                Buffers[bufferIndex] = mutableLastBuffer;
-                int allocatable = (int)Math.Min(count, ReadOnlyDataFrameBuffer<T>.MaxCapacity);
+
+
+                DataFrameBuffer<T> mutableLastBuffer = Buffers.GetMutable(Buffers.Count - 1);
+
+                int allocatable = (int)Math.Min(remaining, ReadOnlyDataFrameBuffer<T>.MaxCapacity);
+
                 mutableLastBuffer.EnsureCapacity(allocatable);
-                mutableLastBuffer.RawSpan.Slice(lastBuffer.Length, allocatable).Fill(value ?? default);
-                mutableLastBuffer.Length += allocatable;
-                Length += allocatable;
+                mutableLastBuffer.Length = allocatable;
+
+                DataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers.GetMutable(NullBitMapBuffers.Count - 1);
+                int nullBufferAllocatable = (allocatable + 7) / 8;
+                lastNullBitMapBuffer.EnsureCapacity(nullBufferAllocatable);
+                lastNullBitMapBuffer.Length = nullBufferAllocatable;
 
-                int nullBitMapBufferIndex = NullBitMapBuffers.Count - 1;
-                ReadOnlyDataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers[nullBitMapBufferIndex];
-                DataFrameBuffer<byte> mutableLastNullBitMapBuffer = DataFrameBuffer<byte>.GetMutableBuffer(lastNullBitMapBuffer);
-                NullBitMapBuffers[nullBitMapBufferIndex] = mutableLastNullBitMapBuffer;
-                int nullBitMapAllocatable = (int)(((uint)allocatable) / 8) + 1;
-                mutableLastNullBitMapBuffer.EnsureCapacity(nullBitMapAllocatable);
-                _modifyNullCountWhileIndexing = false;
-                for (long i = Length - count; i < Length; i++)
+                remaining -= allocatable;
+                Length += mutableLastBuffer.Length;
+
+                // PR Question: Does this need to be called if it's value is null/doesn't have value?
+                if (value.HasValue)
                 {
-                    SetValidityBit(i, value.HasValue ? true : false);
+                    mutableLastBuffer.RawSpan.Slice(mutableLastBuffer.Length, allocatable).Fill(value ?? default);
+
+                    _modifyNullCountWhileIndexing = false;
+                    for (long i = Length - remaining; i < Length; i++)
+                    {
+                        SetValidityBit(i, value.HasValue);
+                    }
+                    _modifyNullCountWhileIndexing = true;
                 }
-                _modifyNullCountWhileIndexing = true;
-                count -= allocatable;
+
             }
         }
 
         public void ApplyElementwise(Func<T?, long, T?> func)
```

Comment:
Out of curiosity, what is the contract for this method? It's going to swap out the buffer in the array with a new one -- what if someone's holding on to the old one? How do you handle reference invalidation?

Reply:
I just inherited this but my understanding is the following:

- This method is going to apply `func ` for each element and update the value in buffer
  - It will make the buffer mutable if it needs to
  - It will update the `nullBitMapBuffer`

If something is holding on to the old buffer - it shouldn't be/it's outdated.